File: packages/server-aspnet/PasskeyConfig.cs
namespace OpenPasskey.AspNet;

/// <summary>
/// Configuration for the passkey endpoints.
/// </summary>
public class PasskeyConfig
{
    public string RpId { get; set; } = "";
    public string RpDisplayName { get; set; } = "";
    public string Origin { get; set; } = "";
    public IChallengeStore ChallengeStore { get; set; } = new MemoryChallengeStore();
    public ICredentialStore CredentialStore { get; set; } = new MemoryCredentialStore();
    public int ChallengeLength { get; set; } = 32;
    public int ChallengeTimeoutSeconds { get; set; } = 300;

    public void Validate()
    {
        if (string.IsNullOrWhiteSpace(RpId))
            throw new ArgumentException("RpId is required");
        if (string.IsNullOrWhiteSpace(Origin))
            throw new ArgumentException("Origin is required");
        if (RpId.Contains("://") || RpId.Contains('/'))
            throw new ArgumentException($"RpId must be a bare domain (got '{RpId}')");
        if (!Origin.StartsWith("https://") && !Origin.StartsWith("http://"))
            throw new ArgumentException($"Origin must start with https:// or http:// (got '{Origin}')");
    }
}

File: packages/server-php/src/PasskeyConfig.php
<?php

declare(strict_types=1);

namespace OpenPasskey\Server;

class PasskeyConfig
{
    public readonly ChallengeStore $challengeStore;
    public readonly CredentialStore $credentialStore;

    public function __construct(
        public readonly string $rpId,
        public readonly string $rpDisplayName,
        public readonly string $origin,
        ?ChallengeStore $challengeStore = null,
        ?CredentialStore $credentialStore = null,
        public readonly int $challengeLength = 32,
        public readonly float $challengeTimeoutSeconds = 300.0,
        public readonly bool $allowMultipleCredentials = false,
        public readonly ?SessionConfig $session = null,
    ) {
        if (str_contains($rpId, '://') || str_contains($rpId, '/')) {
            throw new \InvalidArgumentException('rpId must be a bare domain (no scheme or path)');
        }

        if (!str_starts_with($origin, 'https://') && !str_starts_with($origin, 'http://')) {
            throw new \InvalidArgumentException('origin must start with https:// or http://');
        }

        $this->challengeStore = $challengeStore ?? new MemoryChallengeStore();
        $this->credentialStore = $credentialStore ?? new MemoryCredentialStore();

        if ($session !== null) {
            Session::validateConfig($session);
        }
    }
}
